pkg/client/client.go

```diff
@@ -57,6 +59,15 @@ func New(config *rest.Config, options Options) (Client, error) {
 		return nil, fmt.Errorf("must provide non-nil rest.Config to client.New")
 	}
 
+	// Copy to avoid mutating the passed in config
+	config = rest.CopyConfig(config)
+
+	// Set a WarningHandler
+	config.WarningHandler = rest.NewWarningWriter(os.Stderr, rest.WarningWriterOptions{
```

Comment:
Can we somehow use a logger for this?

Reply:
Hi @alvaroaleman! Thanks for the feedback :smile:

This could work. This implements the `HandleWarningHeader` function as well but [uses](https://github.com/kubernetes/kubernetes/blob/f137c4777095b3972e2dd71a01365d47be459389/staging/src/k8s.io/client-go/rest/warnings.go#L66) `klog` for logging.
```go
config.WarningHandler = rest.WarningLogger{}
```

Reply:
I think I was more thinking about something that would use the logger from `pkg/log`. This might imply a small wrapper?
It might also be a good idea to somehow get the deduplication logic, otherwise we might end up spamming the log

Reply:
Hey @alvaroaleman, I tried thinking about the above suggestion, I'm a little unsure about the following which is why I was hoping to maybe get some feedback on the same:

- For writing a wrapper, did you have something such as the following in mind:
    ```go
    func (l DelegatingLogger) HandleWarningHeader(code int, agent string, text string) {
        ...
    }
    ```
    - And then a `DelegatingLogger` can be assigned as the `config.WarningHandler`
- For deduplication, a new attribute could be added to the `DelegatingLogger` called `written` which can be used to check for duplicate messages similar to as done [here](https://github.com/kubernetes/kubernetes/blob/b0abe89ae259d5e891887414cb0e5f81c969c697/staging/src/k8s.io/client-go/rest/warnings.go#L81).

Reply:
The idea was more to build a wrapper that takes a `logr.Logger` rather than making it part of the `DelegatingLogger` itself in order to keep this composable

Reply:
Oh I see, that makes sense!

To make sure I understand correctly, what you had in mind was a wrapper that is similar to the latest commit, but changed to something that would be used as follows:

```go
rest.SetDefaultWarningHandler(log.NewWarningLogger(
        log.Log.WithName("api-server-warning"),
        log.WarningLoggerOptions{
	        Deduplicate: true,
        }
    )
)
```
Is that what you had in mind?

Reply:
Yes, that looks good. I would probably call it `KubeAPIWarningLogger`, but that is just a detail

Reply:
Awesome! I'll make the changes 😄 

One last question: would it be a good idea to provide a field in [`Options`](https://github.com/kubernetes-sigs/controller-runtime/blob/b704f447ea7c8f7059c6665143a4aa1f6da28328/pkg/client/client.go#L37) to enable/disable surfacing of warnings should there be a case where said warnings aren't desirable?

Reply:
I am agnostic there, I can not think of a reason why someone would not want deduplication but that doesn't mean its impossible.

Reply:
Oh, I meant surfacing of warnings itself and not specifically de-duplication